src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -24,10 +24,298 @@ ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+{
+    ucs_assert((cep->comm_ctx.length != 0));
+    return (cep->comm_ctx.offset == cep->comm_ctx.length);
+}
+
+ucs_status_t uct_tcp_sockcm_ep_progress_send(uct_tcp_sockcm_ep_t *cep,
+                                             int is_async)
+{
+    ucs_status_t status;
+    size_t sent_length;
+
+    ucs_assert(cep->comm_ctx.offset < cep->comm_ctx.length);
+
+    sent_length = cep->comm_ctx.length - cep->comm_ctx.offset;
+
+    status = ucs_socket_send_nb(cep->fd,
+                                UCS_PTR_BYTE_OFFSET(cep->comm_ctx.buf,
+                                                    cep->comm_ctx.offset),
+                                &sent_length, NULL, NULL);
+    if ((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS)) {
+        if (!ucs_socket_is_connected(cep->fd)) {
+            uct_tcp_sockcm_ep_handle_disconnect(cep, status);
+        } else {
+            ucs_error("ep %p failed to send client's data (len=%zu offset=%zu)",
+                      cep, cep->comm_ctx.length, cep->comm_ctx.offset);
+        }
+        return status;
+    }
+
+    cep->comm_ctx.offset += sent_length;
+    ucs_assert(cep->comm_ctx.offset <= cep->comm_ctx.length);
+
+    if (uct_tcp_sockcm_ep_is_tx_rx_done(cep)) {
+        cep->state |= UCT_TCP_SOCKCM_EP_DATA_SENT;
+        uct_tcp_sockcm_ep_init_comm_ctx(cep);
+
+        if (is_async) {
+            /* wait for a reply from the peer */
+            status = ucs_async_modify_handler(cep->fd, UCS_EVENT_SET_EVREAD);
+            if (status != UCS_OK) {
+                ucs_error("failed to modify %d event handler to "
+                          "UCS_EVENT_SET_EVREAD: %s", cep->fd,
+                          ucs_status_string(status));
+            }
+        }
+    } else {
+        cep->state |= UCT_TCP_SOCKCM_EP_SENDING;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_sockcm_ep_send_priv_data(uct_tcp_sockcm_ep_t *cep,
+                                              int is_async)
+{
+    char ifname_str[UCT_DEVICE_NAME_MAX];
+    uct_tcp_sockcm_priv_data_hdr_t *hdr;
+    ssize_t priv_data_ret;
+    ucs_status_t status;
+
+    /* get interface name associated with the connected client fd */
+    status = ucs_sockaddr_get_ifname(cep->fd, ifname_str, UCT_DEVICE_NAME_MAX);
+    if (UCS_OK != status) {
+        goto out;
+    }
+
+    hdr           = (uct_tcp_sockcm_priv_data_hdr_t*)cep->comm_ctx.buf;
+    priv_data_ret = cep->super.priv_pack_cb(cep->super.user_data, ifname_str,
+                                            hdr + 1);
+    if (priv_data_ret < 0) {
+        ucs_assert(priv_data_ret > UCS_ERR_LAST);
+        status = (ucs_status_t)priv_data_ret;
+        ucs_error("tcp_sockcm private data pack function failed with error: %s",
+                  ucs_status_string(status));
+        goto out;
+    } else if (priv_data_ret > (uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len -
+                                sizeof(uct_tcp_sockcm_priv_data_hdr_t))) {
```

Comment:
maybe add function to return `(uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len -                               sizeof(uct_tcp_sockcm_priv_data_hdr_t)`
it's probably also used in sockcm_cm_query

Reply:
if the user is the one that sets the private data length, maybe we should add a comment in this param's doc saying that the actual length that will be used is the value given by him minus ucx's internal header?

also, maybe instead of adding the new func, i'll update the cm's private data len right at the beginning?
https://github.com/openucx/ucx/blob/master/src/uct/tcp/tcp_sockcm.c#L132
self->priv_data_len = cm_config->priv_data_len **- sizeof(uct_tcp_sockcm_priv_data_hdr_t)**;
that way, `uct_tcp_sockcm_ep_get_cm(cep)->priv_data_len` will already be updated to the actual used value and won't need to deduct the header's length everywhere or use the new function

Reply:
ok, no need to over-complicate err report. just update the cm's private data len right at the beginning

Reply:
ok, just update the cm's private data len right at the beginning